server/src/storage/segment.rs

```diff
@@ -0,0 +1,187 @@
+            values: Float32Builder::new(INITIAL_BUILDER_CAPACITY),
+            metadata: data_point.metadata.to_vec(),
+            first_timestamp: data_point.timestamp,
+        }
+    }
+
+    /// Return how many data points the segment currently contains.
+    pub fn get_length(&self) -> usize {
+        // The length is always the same for both builders.
+        self.timestamps.len()
+    }
+
+    /// If at least one of the builders are at capacity, return `true`.
+    pub fn is_full(&self) -> bool {
+        let length = self.get_length();
+        length == self.timestamps.capacity() || length == self.values.capacity()
+    }
+
+    /// Add the timestamp and value from `data_point` to the segment array builders.
+    pub fn insert_data(&mut self, data_point: &DataPoint) {
+        self.timestamps.append_value(data_point.timestamp).unwrap();
+        self.values.append_value(data_point.value).unwrap();
+
+        println!("Inserted data point into {}.", self)
+    }
+
+    /// Finish the array builders and return the data in a structured record batch.
+    pub fn get_data(&mut self) -> RecordBatch {
+        let timestamps = self.timestamps.finish();
+        let values = self.values.finish();
+
+        let schema = Schema::new(vec![
+            Field::new("timestamps", DataType::Timestamp(Microsecond, None), false),
+            Field::new("values", DataType::Float32, false),
+        ]);
+
+        RecordBatch::try_new(
+            Arc::new(schema),
+            vec![Arc::new(timestamps), Arc::new(values)]
+        ).unwrap()
+    }
+
+    /// Write `batch` to persistent parquet file storage.
+    pub fn save_compressed_data(&self, batch: RecordBatch) {
+        let folder_name = self.metadata.join("-");
+        fs::create_dir_all(&folder_name);
+
+        let path = format!("{}/{}.parquet", folder_name, self.first_timestamp);
+        write_batch_to_parquet(batch, path);
+    }
+}
+
+/// Write `batch` to a parquet file at the location given by `path`.
+fn write_batch_to_parquet(batch: RecordBatch, path: String) {
+    // Write the record batch to the parquet file buffer.
+    let file = File::create(path).unwrap();
+    let props = WriterProperties::builder()
+        .set_dictionary_enabled(false)
+        // TODO: Test using more efficient encoding. Plain encoding makes it easier to read the files externally.
+        .set_encoding(Encoding::PLAIN)
+        .build();
+    let mut writer = ArrowWriter::try_new(file, batch.schema(), Some(props)).unwrap();
+
+    writer.write(&batch).expect("Writing batch.");
+    writer.close().unwrap();
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use paho_mqtt::Message;
+
+    fn get_empty_segment_builder() -> (DataPoint, SegmentBuilder) {
+        let message = Message::new("ModelarDB/test", "[1657878396943245, 30]", 1);
+
+        let data_point = DataPoint::from_message(&message).unwrap();
+        let segment_builder = SegmentBuilder::new(&data_point);
```

Comment:
Seems a bit unintuitive that `get_empty_segment_builder()` constructs a `SegmentBuilder` that contains a single `DataPoint` but has length zero.

Reply:
This is a quirk of the current design and is no longer a problem in the next branch where `SegmentBuilder` can be created without a data point.